parser/src/actions.ts

```diff
@@ -0,0 +1,310 @@
+import * as t from "@babel/types";
+import { Action, ChooseCondition } from "xstate";
+import { assign, choose, forwardTo, send } from "xstate/lib/actions";
+import { Cond, CondNode } from "./conds";
+import { createParser } from "./createParser";
+import { maybeIdentifierTo } from "./identifiers";
+import {
+  AfterAction,
+  CancelAction,
+  DoneAction,
+  EscalateAction,
+  LogAction,
+  PureAction,
+  RaiseAction,
+  RespondAction,
+  SendParentAction,
+  SendUpdateAction,
+  StartAction,
+  StopAction,
+} from "./namedActions";
+import { AnyNode, NumericLiteral, StringLiteral } from "./scalars";
+import { maybeTsAsExpression } from "./tsAsExpression";
+import { DeclarationType } from "./types";
+import { unionType } from "./unionType";
+import {
+  arrayOf,
+  isFunctionOrArrowFunctionExpression,
+  maybeArrayOf,
+  namedFunctionCall,
+  objectTypeWithKnownKeys,
+} from "./utils";
+import { wrapParserResult } from "./wrapParserResult";
+
+export interface ActionNode {
+  node: t.Node;
+  action: Action<any, any>;
+  name: string;
+  chooseConditions?: ParsedChooseCondition[];
+  declarationType: DeclarationType;
+  inlineDeclarationId: string;
+}
+
+export interface ParsedChooseCondition {
+  condition: ChooseCondition<any, any>;
+  actionNodes: ActionNode[];
+  conditionNode?: CondNode;
+}
+
+export const ActionAsIdentifier = maybeTsAsExpression(
+  createParser({
+    babelMatcher: t.isIdentifier,
+    parseNode: (node, context): ActionNode => {
+      return {
+        action: node.name,
+        node,
+        name: node.name,
+        declarationType: "identifier",
+        inlineDeclarationId: context.getNodeHash(node),
+      };
+    },
+  }),
+);
+
+export const ActionAsFunctionExpression = maybeTsAsExpression(
+  maybeIdentifierTo(
+    createParser({
+      babelMatcher: isFunctionOrArrowFunctionExpression,
+      parseNode: (node, context): ActionNode => {
+        const action = function actions() {};
```

Comment:
q: why do we need this? it's a dummy value so technically shouldn't be needed, right? couldn't the `action` property be optional? or part of a discriminated union like `NamedAction | InlineAction`?

Reply:
It's a slight hangover from the way I've been testing this:

https://www.loom.com/share/bf8a1d808e8743d3b7f25ed97fa49c1f